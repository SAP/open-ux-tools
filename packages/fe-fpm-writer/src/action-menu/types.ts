import type { CustomElement, Position } from '../common/types';

export enum TargetControl {
    header = 'header',
    body = 'body',
    //    footer = 'footer',
    section = '@com.sap.vocabularies.UI.v1.FieldGroup',
    table = '@com.sap.vocabularies.UI.v1.LineItem'
}

export interface ActionMenuTarget {
    page: string;
    control: TargetControl;
    customSectionKey?: string;
    navProperty?: string;
    qualifier?: string;
}

export interface ActionMenu extends CustomElement {
    target: ActionMenuTarget;
    settings: {
        text: string;
        actions: string[];
        position?: Position;
    };
}
