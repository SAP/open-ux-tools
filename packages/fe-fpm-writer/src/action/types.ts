import type { CustomElement, InternalCustomElement, Position, EventHandler } from '../common/types';

export enum TargetControl {
    header = 'header',
    footer = 'footer',
    section = '@com.sap.vocabularies.UI.v1.FieldGroup',
    table = '@com.sap.vocabularies.UI.v1.LineItem'
}

export interface CustomActionTarget {
    page: string;
    control: TargetControl;
    navProperty?: string;
    qualifier?: string;
}

export interface CustomAction extends CustomElement, EventHandler {
    target: CustomActionTarget;
    settings: {
        text: string;
        visible?: string | true;
        enabled?: string | true;
        requiresSelection?: boolean;
        position?: Position;
    };
}

export type InternalCustomAction = CustomAction & InternalCustomElement;
