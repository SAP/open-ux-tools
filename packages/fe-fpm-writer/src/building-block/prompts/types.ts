import type { Answers as IAnswers } from 'inquirer';
import type { PromptQuestion } from '@sap-ux/ui-prompting';
import type { BuildingBlockType, Chart, FilterBar, Table } from '../types';

export interface PromptsGroup {
    id: string;
    title: string;
    description: string[];
}

export type Answers = IAnswers;

export interface Prompts<T extends Answers> {
    questions: PromptQuestion<T>[];
    groups?: PromptsGroup[];
}

export interface BuildingBlockTypePromptsAnswer extends Answers {
    buildingBlockType: BuildingBlockType;
}

export interface TablePromptsAnswer extends Table, Answers {
    viewOrFragmentFile: string;
    aggregationPath: string;
    id: string;
    entity: string;
    filterBar: string;
    selectionChange: string;
    bindingContextType: 'relative' | 'absolute';
    qualifier: string;
    type: 'ResponsiveTable' | 'GridTable';
    displayHeader: boolean;
    tableHeaderText: string;
}

export interface ChartPromptsAnswer extends Chart, Answers {
    viewOrFragmentFile: string;
    aggregationPath: string;
    id: string;
    entity: string;
    filterBar: string;
    selectionMode: string;
    selectionChange: string;
    qualifier: string;
    bindingContextType: 'relative' | 'absolute';
}

export interface FilterBarPromptsAnswer extends FilterBar, Answers {
    qualifier: string;
    entity: string;
    viewOrFragmentFile: string;
}

// ToDo - import from ui types
export type ValidationResults = { [questionName: string]: { isValid: boolean; errorMessage?: string } };

export interface AdditionalPromptProperties {
    required?: boolean;
    groupId?: string;
    additionalInfo?: string;
    placeholder?: string;
}
