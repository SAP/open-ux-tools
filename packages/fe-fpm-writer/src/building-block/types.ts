/**
 * Building block type.
 *
 * @enum {string}
 */
export enum BuildingBlockType {
    FilterBar = 'filter-bar',
    Chart = 'chart',
    Field = 'field'
}

/**
 * Represents a building block control.
 */
export interface BuildingBlock {
    /**
     * Building block type.
     */
    buildingBlockType: BuildingBlockType;

    /**
     * Defines the path of the context used in the current page or block. This setting is defined by the framework.
     */
    contextPath?: string;

    /**
     * ID of the building block.
     */
    id: string;

    /**
     * Defines the relative path of the property in the metamodel, based on the current contextPath.
     */
    metaPath?: string;
}

/**
 * Represents a Chart building block control.
 * Usually, a contextPath and metaPath is expected.
 *
 * @example
 * <macro:Chart id="Mychart" contextPath="/RootEntity" metaPath="@com.sap.vocabularies.UI.v1.Chart" />
 * @extends {BuildingBlock}
 */
export interface Chart extends BuildingBlock {
    /**
     * ID of the FilterBar building block associated with the chart.
     */
    filterBar?: string;
    /**
     * Sets the personalization of the chart
     */
    personalization?: boolean | string;
    /**
     * Specifies the selection mode.
     */
    selectionMode?: string;
    /**
     * An event triggered when chart selections are changed.
     * The event contains information about the data selected/deselected and the Boolean flag that indicates
     * whether data is selected or deselected.
     */
    selectionChange?: string;
}

/**
 * Represents a custom filter to be used inside the FilterBar.
 * The template for the FilterField has to be provided as the default aggregation.
 */
export interface FilterField {
    /**
     * Reference to the key of another filter already displayed in the table to properly place this one.
     */
    anchor: string;
    /**
     * The property name of the FilterField.
     */
    key: string;
    /**
     * The text that will be displayed for this FilterField.
     */
    label: string;
    /**
     * If set the search will be automatically triggered, when a filter value was changed.
     */
    liveMode: boolean;
    /**
     * Defines where this filter should be placed relative to the defined anchor.
     * Allowed values are `Before` and `After`.
     */
    placement: 'Before' | 'After';
    /**
     * Parameter which sets the visibility of the FilterBar building block.
     */
    visible: boolean;
}

/**
 * Represents a Filter Bar building block control.
 * Usually, a SelectionFields annotation is expected.
 *
 * @example
 * <macro:FilterBar id="MyFilterBar" metaPath="@com.sap.vocabularies.UI.v1.SelectionFields" />
 * @extends {BuildingBlock}
 */
export interface FilterBar extends BuildingBlock {
    /**
     * This event is fired after either a filter value or the visibility of a filter item has been changed.
     * The event contains conditions that will be used as filters.
     */
    filterChanged?: string;
    /**
     * This event is fired when the Go button is pressed or after a condition change.
     */
    search?: string;
}

/**
 * Represents the format options for a field control.
 */
export interface FieldFormatOptions {
    /**
     * Defines how the field value and associated text will be displayed together.
     * Allowed values are "Value", "Description", "ValueDescription" and "DescriptionValue"
     */
    displayMode?: 'Value' | 'Description' | 'ValueDescription' | 'DescriptionValue';
    /**
     * Defines if and how the field measure will be displayed.
     * Allowed values are "Hidden" and "ReadOnly"
     */
    measureDisplayMode?: 'Hidden' | 'ReadOnly';
    /**
     * Defines how the full text will be displayed.
     * Allowed values are "InPlace" and "Popover"
     */
    textExpandBehaviorDisplay?: 'InPlace' | 'Popover';
    /**
     * Maximum number of lines for multiline texts in edit mode.
     */
    textLinesEdit?: number;
    /**
     * Maximum number of characters from the beginning of the text field that are shown initially.
     */
    textMaxCharactersDisplay?: number;
    /**
     * Maximum number of lines that multiline texts in edit mode can grow to.
     */
    textMaxLines?: number;
}

/**
 * Building block for creating a field based on the metadata provided by OData V4.
 * Usually, a DataField or DataPoint annotation is expected, but the field can also be used to display
 * a property from the entity type.
 *
 * @example
 * <macro:Field id="MyField" metaPath="MyProperty" />
 * @extends {BuildingBlock}
 */
export interface Field extends BuildingBlock {
    /**
     * A set of format options that can be configured.
     */
    formatOptions?: FieldFormatOptions;
    /**
     * An expression that allows you to control the read-only state of the field.
     * If you do not set any expression, SAP Fiori elements hooks into the standard lifecycle
     * to determine the current state.
     */
    readOnly?: boolean;
    /**
     * Option to add semantic objects to a field.
     * Valid options are either a single semantic object, a stringified array of semantic objects
     * or a single binding expression returning either a single semantic object or an array of semantic objects
     */
    semanticObject?: string;
}

/**
 * Building block for creating a Form based on the metadata provided by OData V4.
 * It is designed to work based on a FieldGroup annotation but can also work if you provide a
 * ReferenceFacet or a CollectionFacet.
 *
 * @example
 * <macro:Form id="MyForm" metaPath="@com.sap.vocabularies.UI.v1.FieldGroup#GeneralInformation" />
 * @extends {BuildingBlock}
 */
export interface Form extends BuildingBlock {
    /**
     * The title of the form control.
     */
    title?: string;
}

/**
 * Building block used to create a form element containing a label and a field.
 *
 * @extends {BuildingBlock}
 */
export interface FormElement extends BuildingBlock {
    /**
     * Label shown for the field. If not set, the label from the annotations will be shown.
     */
    label: string;
    /**
     * If set to false, the FormElement is not rendered.
     */
    visible: boolean;
}

/**
 * Building block used to create a micro chart based on the metadata provided by OData V4.
 *
 * @extends {BuildingBlock}
 */
export interface MicroChart extends BuildingBlock {
    /**
     * Batch group ID along with which this call should be grouped.
     */
    batchGroupId?: string;
    /**
     * Show blank space in case there is no data in the chart.
     */
    hideOnNoData?: boolean;
    /**
     * To control the rendering of Title, Subtitle and Currency Labels. When the size is xs then we do not see
     * the inner labels of the micro chart as well.
     */
    showOnlyChart?: boolean;
    /**
     * Size of the micro chart control.
     */
    size?: string;
}

/**
 * Building block used to create a paginator control.
 *
 * @example
 * <macro:Paginator id="someID" />
 * @extends {BuildingBlock}
 */
export type Paginator = BuildingBlock;

/**
 * Building block used to create the share functionality.
 * Please note that the 'Share in SAP Jam' option is only available on platforms that are integrated with SAP Jam.
 * If you are consuming this macro in an environment where the SAP Fiori launchpad is not available,
 * then the 'Save as Tile' option is not visible.
 *
 * @example
 * <macro:Share id="someID" visible="true" />
 * @extends {BuildingBlock}
 */
export interface Share extends BuildingBlock {
    /**
     * If set to false, the share control is not rendered.
     */
    visible: boolean;
}

/**
 * Building block used to create a table based on the metadata provided by OData V4.
 * Usually, a LineItem or PresentationVariant annotation is expected,
 * but the table building block can also be used to display an EntitySet.
 *
 * @example
 * <macro:Table id="MyTable" metaPath="@com.sap.vocabularies.UI.v1.LineItem" />
 * @extends {BuildingBlock}
 */
export interface Table extends BuildingBlock {
    /**
     * An expression that allows you to control the 'busy' state of the table.
     */
    busy?: boolean;
    /**
     * Specifies the header text that is shown in the table.
     */
    enableAutoColumnWidth?: boolean;
    /**
     * Controls if the export functionality of the table is enabled or not.
     * Default value is true.
     */
    enableExport?: boolean;
    /**
     * Controls whether the table can be opened in fullscreen mode or not.
     */
    enableFullScreen?: boolean;
    /**
     * Controls if the paste functionality of the table is enabled or not.
     */
    enablePaste?: boolean;
    /**
     * ID of the FilterBar building block associated with the table.
     */
    filterBar?: string;
    /**
     * Specifies the header text that is shown in the table.
     */
    header?: string;
    /**
     * Controls if the header text should be shown or not.
     * Default value is true.
     */
    headerVisible?: boolean;
    /**
     * Defines whether to display the search action.
     */
    isSearchable?: boolean;
    /**
     * Groups menu actions by key.
     */
    menu?: string;
    /**
     * Controls which options should be enabled for the table personalization dialog.
     * If it is set to `true`, all possible options for this kind of table are enabled.
     * If it is set to `false`, personalization is disabled.
     * You can also provide a more granular control for the personalization by providing a comma-separated list
     * with the options you want to be available.
     * Available options are: `Sort`, `Column` and `Filter`
     */
    personalization?: boolean | 'Sort' | 'Column' | 'Filter';
    /**
     * An expression that allows you to control the 'read-only' state of the table.
     * If you do not set any expression, SAP Fiori elements hooks into the standard lifecycle to determine
     * the current state.
     */
    readOnly?: boolean;
    /**
     * Defines the selection mode to be used by the table.
     * Allowed values are `None`, `Single`, `Multi` or `Auto`
     */
    selectionMode?: 'None' | 'Single' | 'Multi' | 'Auto';
    /**
     * Defines the type of table that will be used by the building block to render the data.
     * Allowed values are `GridTable` and `ResponsiveTable`.
     * Default value is `ResponsiveTable`.
     */
    type?: 'GridTable' | 'ResponsiveTable';
    /**
     * Controls the kind of variant management that should be enabled for the table.
     * Allowed values are `Page`, `Control` and `None`.
     * If the table is used within a SAP Fiori elements template, the default value will be taken from
     * the current page variant management.
     * Otherwise it's `None`.
     */
    variantManagement?: string;
}

/**
 * Input configuration for the generate function.
 */
export interface BuildingBlockConfig<T extends BuildingBlock> {
    /**
     * The path of the view or fragment xml file relative to the base path.
     */
    viewOrFragmentPath: string;

    /**
     * The aggregation xpath.
     */
    aggregationPath: string;

    /**
     * The building block parameters.
     */
    buildingBlockData: T;
}
