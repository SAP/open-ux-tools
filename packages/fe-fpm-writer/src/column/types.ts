import type { CustomElement, InternalCustomElement, Position, EventHandler } from '../common/types';

export enum Availability {
    'Default' = 'Default',
    'Adaptation' = 'Adaptation',
    'Hidden' = 'Hidden'
}

export enum HorizontalAlign {
    'Begin' = 'Begin',
    'Center' = 'Center',
    'End' = 'End'
}

export type ColumnPropertiesType = string[];

/**
 * Custom Column.
 *
 */
export interface CustomTableColumn extends CustomElement, EventHandler {
    /**
     * Name of the routing target
     */
    target: string;
    /**
     * Name of the target entity for the control configuration
     */
    targetEntity: string;
    /**
     *   Defines the position of the column relative to other columns.
     */
    position: Position;
    /**
     * The header is shown on the table as header, as well as in the add/remove dialog.
     */
    header: string;
    /**
     * Optional control XML that will be generated into the fragment of table column. If the property isn't provided then a sample control will be generated.
     */
    control?: string;
    /**
     * A string type that represents CSS size values.
     * Refer to https://openui5.hana.ondemand.com/api/sap.ui.core.CSSSize.
     */
    width?: string;
    /**
     * Aligns the header as well as the content horizontally.
     */
    horizontalAlign?: HorizontalAlign;
    /**
     * Defines where the column should be shown.
     * - Default: it will be shown by default in the table.
     * - Adaptation: it will initially not shown in the table but be available via end user adaptation
     * - Hidden: the column is neither available in the table nor in adaptation
     */
    availability?: Availability;
    /**
     * If provided and sorting is enabled for the table the custom column header can be clicked. This will lead to a list of properties (or one) that can be sorted by, they are displayed as the labels of the properties, corresponding to their definition in the annotations.
     */
    properties?: ColumnPropertiesType;
}

export interface InternalCustomTableColumn extends CustomTableColumn, InternalCustomElement {
    content: string;
}
