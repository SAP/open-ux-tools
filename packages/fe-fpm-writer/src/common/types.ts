import type { ManifestNamespace } from '@sap-ux/project-access';

export type Manifest = ManifestNamespace.SAPJSONSchemaForWebApplicationManifestFile & { [key: string]: unknown };

/**
 * Common properties for any custom element of the flexible programming model.
 */
export interface CustomElement {
    /**
     * Name of the custom element that is to be added to the application.
     */
    name: string;
    /**
     * Target folder relative to the manifest that is used to generate the custom extension content (e.g. view, fragment, controller).
     * If undefined then `ext/${id}` is used.
     */
    folder?: string;
    /**
     * Optional property to define the minimum UI5 version that the generated code must support.
     * If undefined then the latest version of the template is used.
     * If nothing can be generated for the given version then an exception is thrown.
     */
    minUI5Version?: string;
    /**
     * If set to true then all code snippets are generated in Typescript instead of Javascript.
     */
    typescript?: boolean;
}

export interface InternalCustomElement extends CustomElement {
    ns: string;
    path: string;
}

/**
 * Enumeration with all values possible as position placement attribute
 */
export enum Placement {
    After = 'After',
    Before = 'Before',
    End = 'End'
}

/**
 * Position of a custom element relative to an anchor element.
 */
export type Position = {
    /**
     * The key of another element to be used as placement anchor.
     */
    anchor?: string;
    /**
     * Define the placement, either before or after the anchor element or at the end.
     */
    placement: Placement;
};

export interface Ui5RoutingTarget<T> extends ManifestNamespace.Target {
    options?: T;
}

export interface ViewVariant {
    /**
     * The key entry is used for initializing the corresponding IconTabBar item.
     */
    key: string;
    /**
     * The annotationPath of a specific variant.
     */
    annotationPath: string;
    /**
     * View title, is shown in the tab item.
     */
    label?: string;
    /**
     * If you switch from an app with only a single table to an app with multiple tables, you can keep the previously defined variant on one tab.
     */
    keepPreviousPersonalization?: boolean;
    /**
     * The template being used for the custom view.
     */
    template?: string;
}

export interface Views {
    /**
     * The paths section contains a set of entries that point to the variants defined in the annotations.
     */
    paths?: Array<ViewVariant>;
    /**
     * Determines whether the count is displayed in the tabs (default setting: false).
     */
    showCounts?: boolean;
}

export interface Ui5TargetSettings {
    settings?: {
        /**
         * Represents the entity set that defines either the aggregation or the root object of the component.
         */
        entitySet?: string;
        /**
         * By default, the list report displays only one table. You can define multiple views of a table, and add a chart, if required.
         */
        views?: Views;
    };
}

/**
 * Represents a line and character position in file.
 */
export interface FileContentPosition {
    /**
     * Zero based line index.
     */
    line: number;
    /**
     * Zero based character value.
     */
    character: number;
}

/**
 * Interface for text fragment insertion into existing text file.
 */
export interface TextFragmentInsertion {
    /**
     * Insert position for passed script fragment.
     * Position can be line + char indices or absolute char index
     */
    position: FileContentPosition | number;
    /**
     * Fragment of script to insert.
     */
    fragment: string;
}

export interface EventHandlerConfiguration {
    /**
     * JS file name - existing or new.
     */
    fileName?: string;
    /**
     * Function name. If undefined then `onPress` is used as default value.
     */
    fnName?: string;
    /**
     * If file exists, then existing file should be appended with passed script fragment.
     */
    insertScript?: TextFragmentInsertion;
}

export interface EventHandler {
    /**
     * If not set (i.e. undefined) then no event handler is linked. If it is set true, a new one is created and linked to the action.
     * If an existing event handler is to be used then its id needs to be provided as string.
     * Object with 'EventHandlerConfiguration' provides following options:
     * 1. Append existing js file with new action by providing script fragment;
     * 2. Pass non-default function name(default is "onPress");
     */
    eventHandler?: true | string | EventHandlerConfiguration;
}
