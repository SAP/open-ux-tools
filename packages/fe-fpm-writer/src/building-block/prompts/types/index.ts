export * from './basic';
export * from './prompts';
