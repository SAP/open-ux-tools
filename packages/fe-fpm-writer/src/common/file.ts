import type { Editor } from 'mem-fs-editor';
import type { TabInfo } from '../common/types';

const CHAR_SPACE = ' ';
const CHAR_TAB = '\t';

type WriteJsonReplacer = ((key: string, value: any) => any) | Array<string | number>;

type WriteJsonSpace = number | string;
interface ExtendJsonParams {
    filepath: string;
    content: string;
    replacer?: WriteJsonReplacer;
    tabInfo?: TabInfo;
}

/**
 * Method returns tab info for passed line.
 *
 * @param line - line with tab spacing
 * @returns tab size information
 */
function getLineTabInfo(line: string): TabInfo | undefined {
    let tabSize: TabInfo | undefined;
    const symbol = line[0] === CHAR_TAB ? CHAR_TAB : CHAR_SPACE;
    // get count of tabs
    for (let i = 0; i < line.length; i++) {
        const char = line[i];
        if (char !== symbol) {
            tabSize = {
                size: i,
                useTabSymbol: symbol === CHAR_TAB
            };
            break;
        }
    }
    return tabSize;
}

/**
 * Method calculates tab space info for passed file content.
 *
 * @param content - file content.
 * @returns tab size information.
 */
export function detectTabSpacing(content: string): TabInfo | undefined {
    let tabSize: TabInfo | undefined;
    const tabSymbols = [CHAR_SPACE, CHAR_TAB];
    const lines = content.split(/\r\n|\n/);
    const lineWithSpacing = lines.find((line: string): boolean => {
        return tabSymbols.includes(line[0]);
    });
    if (lineWithSpacing) {
        tabSize = getLineTabInfo(lineWithSpacing);
    }
    return tabSize;
}

/**
 * Method calculates tab spacing parameter for 'JSON.stringify' method.
 *
 * @param fs - the mem-fs editor instance.
 * @param filePath - path to file to read.
 * @param tabInfo - External tab configuration.
 * @returns tab size information.
 */
export function getJsonSpace(fs: Editor, filePath: string, tabInfo?: TabInfo | undefined): WriteJsonSpace | undefined {
    if (!tabInfo) {
        // 'tabInfo'  was not passed - calculate 'tabInfo' by checking existing content of target file
        const content = fs.read(filePath);
        tabInfo = detectTabSpacing(content);
    }
    let space: WriteJsonSpace | undefined;
    if (tabInfo) {
        // 'tabInfo' exists - it was passed as custom configuration or calculated from target file
        if (tabInfo.useTabSymbol) {
            // Tab symbol should be used as tab
            space = CHAR_TAB.repeat(tabInfo.size || 1);
        } else {
            // Spaces should be used as tab
            space = tabInfo.size;
        }
    }
    return space;
}

/**
 * Method extends target JSON file with passed JSOn content.
 * Method uses 'fs.extendJSON', but applies additional calculation to reuse existing content tab sizing information.
 *
 * @param fs - the mem-fs editor instance.
 * @param params - options for JSON extend.
 */
export function extendJSON(fs: Editor, params: ExtendJsonParams): void {
    const { filepath, content, replacer } = params;
    const space: WriteJsonSpace | undefined = getJsonSpace(fs, filepath, params.tabInfo);
    // Write json
    fs.extendJSON(filepath, JSON.parse(content), replacer, space);
}
