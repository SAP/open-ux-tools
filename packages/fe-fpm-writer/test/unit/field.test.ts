import type { Editor } from 'mem-fs-editor';
import { create } from 'mem-fs-editor';
import { create as createStorage } from 'mem-fs';
import { join } from 'node:path';
import { generateCustomField } from '../../src/field';
import type { CustomField } from '../../src/field/types';
import type { EventHandlerConfiguration, Manifest } from '../../src/common/types';
import { Placement } from '../../src/common/types';
import * as manifest from './sample/field/webapp/manifest.json';
import { detectTabSpacing } from '../../src/common/file';
import { getEndOfLinesLength, tabSizingTestCases } from '../common';

const testDir = join(__dirname, 'sample/field');

describe('CustomField', () => {
    let fs: Editor;
    const customField: CustomField = {
        target: 'sample',
        targetEntity: '@com.sap.vocabularies.UI.v1.FieldGroup',
        name: 'NewCustomField',
        folder: 'extensions/custom',
        label: 'New Custom Field',
        position: {
            placement: Placement.After,
            anchor: 'DummyField'
        }
    };

    const expectedFragmentPath = join(testDir, `webapp/${customField.folder}/${customField.name}.fragment.xml`);
    beforeEach(() => {
        fs = create(createStorage());
        fs.delete(testDir);
        fs.write(join(testDir, 'webapp/manifest.json'), JSON.stringify(manifest));
    });

    test('with fragmentFile', async () => {
        const testCustomField: CustomField = {
            ...customField,
            fragmentFile: 'NewCustomFieldFragment'
        };
        const expectedFieldFragmentPath = join(
            testDir,
            `webapp/${testCustomField.folder}/${testCustomField.fragmentFile}.fragment.xml`
        );
        await generateCustomField(testDir, { ...testCustomField }, fs);
        const updatedManifest = fs.readJSON(join(testDir, 'webapp/manifest.json')) as Manifest;
        const settings = (
            updatedManifest['sap.ui5']?.['routing']?.['targets']?.['sample']?.['options'] as Record<string, any>
        )['settings'];
        expect(settings.controlConfiguration).toMatchSnapshot();
        expect(fs.read(expectedFieldFragmentPath)).toMatchSnapshot();
    });
});
