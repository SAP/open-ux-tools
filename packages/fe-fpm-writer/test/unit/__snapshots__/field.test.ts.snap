// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CustomField with fragmentFile 1`] = `
Object {
  "@com.sap.vocabularies.UI.v1.FieldGroup": Object {
    "fields": Object {
      "ExistingCustomField": Object {
        "label": "{i18n>customField}",
        "position": Object {
          "anchor": "DataField::BooleanProperty",
          "placement": "After",
        },
        "template": "sap.fe.core.fpmExplorer.customFieldContent.CustomFieldText",
      },
      "NewCustomField": Object {
        "label": "New Custom Field",
        "position": Object {
          "anchor": "DummyField",
          "placement": "After",
        },
        "template": "sapux.fe.fpm.writer.test.extensions.custom.NewCustomFieldFragment",
      },
    },
  },
}
`;

exports[`CustomField with fragmentFile 2`] = `
"<core:FragmentDefinition xmlns:core=\\"sap.ui.core\\" xmlns=\\"sap.m\\">
	<Text text=\\"NewCustomField\\" />
</core:FragmentDefinition>"
`;
